src/libraries/System.Text.Json/tests/JsonNode/JsonObjectTests.cs

```diff
@@ -0,0 +1,515 @@
+
+            Assert.Equal(1, jObject.Keys.Count);
+            Assert.Equal(1, jObject.Values.Count);
+        }
+
+        [Fact]
+        public static void Clear_ContainsKey()
+        {
+            var jObject = new JsonObject();
+            jObject.Add("One", 1);
+            jObject.Add("Two", 2);
+            Assert.Equal(2, jObject.Count);
+            JsonNode node1 = jObject["One"];
+            JsonNode node2 = jObject["Two"];
+
+            Assert.True(jObject.ContainsKey("One"));
+            Assert.True(jObject.ContainsKey("Two"));
+            Assert.False(jObject.ContainsKey("?"));
+
+            jObject.Clear();
+            Assert.Equal(0, jObject.Count);
+
+            jObject.Add("One", 1);
+            jObject.Add("Two", 2);
+            Assert.Equal(2, jObject.Count);
+        }
+
+        [Fact]
+        public static void CaseSensitivity_ReadMode()
+        {
+            var options = new JsonSerializerOptions();
+            JsonObject obj = JsonSerializer.Deserialize<JsonObject>("{\"MyProperty\":42}", options);
+
+            Assert.Equal(42, obj["MyProperty"].GetValue<int>());
+            Assert.Null(obj["myproperty"]);
+            Assert.Null(obj["MYPROPERTY"]);
+
+            options = new JsonSerializerOptions();
+            options.PropertyNameCaseInsensitive = true;
+            obj = JsonSerializer.Deserialize<JsonObject>("{\"MyProperty\":42}", options);
+
+            Assert.Equal(42, obj["MyProperty"].GetValue<int>());
+            Assert.Equal(42, obj["myproperty"].GetValue<int>());
+            Assert.Equal(42, obj["MYPROPERTY"].GetValue<int>());
+        }
+
+        [Fact]
+        public static void CaseSensitivity_EditMode()
+        {
+            var jArray = new JsonArray();
+            var jObject = new JsonObject();
+            jObject.Add("MyProperty", 42);
+            jObject.Add("myproperty", 42); // No exception
+
+            // Options on direct node.
+            var options = new JsonNodeOptions { PropertyNameCaseInsensitive = true };
+            jArray = new JsonArray();
+            jObject = new JsonObject(options);
+            jObject.Add("MyProperty", 42);
+            jArray.Add(jObject);
+            Assert.Throws<ArgumentException>(() => jObject.Add("myproperty", 42));
+
+            // Options on parent node.
+            jArray = new JsonArray(options);
+            jObject = new JsonObject();
+            jArray.Add(jObject);
+            jObject.Add("MyProperty", 42);
+            Assert.Throws<ArgumentException>(() => jObject.Add("myproperty", 42));
+
+            // Dictionary is created when Add is called for the first time, so we need to be added first.
+            jArray = new JsonArray(options);
+            jObject = new JsonObject();
+            jObject.Add("MyProperty", 42);
+            jArray.Add(jObject);
+            jObject.Add("myproperty", 42); // no exception since options were not set in time.
```

Comment:
With collection initializers this can occur although I don't see an easy alternative to "force" the options to be specified upfront before "Add()".

Originally, I had a design where the `JsonSerializerOptions` were used instead of `JsonNodeOptions` since those options have a concept of case insensitivity. However, that dependency was removed because tying those options to DOM creation was somewhat confusing since basically only the case insensitivity flag was used and nothing else. However, using `JsonSerializerOptions` like originally proposed could work in this scenario when we implement the ability to change the global options (as is scheduled for v6 I believe).

So if we add the ability to change the global `JsonSerializerOptions` perhaps we also add the ability to change\specify a global `JsonNodeOptions` and that could be used to address this scenario.

Reply:
I see - yeah we should document this explicitly. 

> when we implement the ability to change the global options (as is scheduled for v6 I believe).

I'm getting more and more scared of doing this due to side effects with the defaults being changed in a way that affects all assemblies in a compilation.